# SNPCall_Benchmarking
Repo for scripts and files used in benchmarking variant callers on simulated shotgun metagenomic data.

# Methods & Workflow

<br/><br/>

<p align="center">
    <img src=https://github.com/TheNoyesLab/SNPCall_Benchmarking/blob/main/Benchmark_Workflow.png>
</p>

<p align="center">
    <b>Figure 1.</b> Workflow diagram showing the variant caller benchmarking process. First, select RefSeq genomes were chosen to simulate a metagenome, random mutations were added to the genomes to create a "gold standard" dataset, then the number of genomes used and number of reads created were adjusted to evaluate the variant callers under a range of sample conditions.
</p>

<br/><br/>


# Results

<br/><br/>

<p align="center">
    <img src=https://github.com/TheNoyesLab/SNPCall_Benchmarking/blob/main/AccuracyBenchmarks.png>
</p>

<p align="center">
    <img src=https://github.com/TheNoyesLab/SNPCall_Benchmarking/blob/main/MemoryBenchmarks.png>
</p>

<br/><br/>


# Important Directories
## projects/SNP\_Call\_Benchmarking/*Benchmarking\_Run*:
- Directory containing synthetic reads, variant caller output, and benchmarks
- All output of Benchmarking Workflow goes here

## SNPCall_Benchmarking/**Workflow**:
Production-stage scripts that form the core Benchmarking workflow.
### Genesis.sh
  - Script performing directory setup, SNP generation, read synthesis, and alignment

### MultiCaller.sh
  - Script running variant callers and benchmarking on data generated by Genesis.sh

### Single_scripts/
  - Core workflow broken up into individual scripts (read generation,alignment,individual variant callers, etc)

### SNP_Injector_Fasta.py
  - Python code in Genesis.sh used to "inject" SNPs into fasta files
  - Creates a log of input SNPs and genome locations

